docs/proposals/config_data.md

```diff
@@ -0,0 +1,319 @@
+
+### Backwards-Compatibility
+In order to smooth the transaction of component configuration from other means to ``ConfigData``
+and preserve backwards-compatibility, user will be able to configure components the old way
+and use an optional ``--write-config=<path>`` flag to create a ``ConfigData`` object containing
+the given configuration options to ``<path>``, which can be created by invoking ``kubectl create -f <path>``.
+
+For example, a ``kubelet`` will be configured as follows:
+```bash
+kubelet --address=0.0.0.0 --port=8888 --api-servers=10.240.13.14:8998,10.240.51.5:8998 --register-node=true --write-file=kubelet-config.yaml
+```
+
+The persisted object will have the following contents for the given configuration parameters:
+```json
+{
+    "apiVersion": "v1",
+    "kind": "ConfigData",
+    "data": {
+        "ADDRESS": "0.0.0.0",
+        "PORT": "8888",
+        "API_SERVERS": "10.240.13.14:8998, 10.240.51.5:8998",
+        "REGISTER_NODE": "true",
+    }
+}
+```
+
+An identically configured kubelet can be created as such from now on:
+```bash
+kubelet --config-data=kubelet-config.yaml
+```
+
+The command-line flags given to the ``kubelet`` will be transformed to ``Data`` fields
+inside a new ``ConfigData`` object for its configuration.
+
+**Note:** Tooling will be implemented to facilitate this functionality for each
+component that is going to migrate to ``ConfigData``.
+
+### Explicit Creation
+The following file ``pod-config.json`` contains data intended to be consumed as environment variables.
+```json
+{
+    "apiVersion": "v1",
+    "name": "PodConfiguration",
+    "kind": "ConfigData",
+    "data": {
+        "server_address": "10.240.13.14:8998",
+        "db_address": "10.240.55.3",
+        "db_port": "3306"
+    }
+}
+```
+
+To persist this ConfigData object to the API server, use the following command:
+```bash
+kubectl create -f pod-config.json
+```
+
+Now, the ``PodConfiguration`` ConfigData object is stored in ``etcd``.
+The new object can be accessed, referenced, used and updated through the API server.
+
+## Configuration Workflow
+
+The component configuration workflow is given below:
+
+1. ``ConfigData`` object is created through the API server.
+2. New components using the ``ConfigData`` object are created.
+3. ``ConfigData`` object is retrieved using the downward API.
+4. ``Data`` key/value pairs of ``ConfigData`` will be exposed as either environment variables (``EnvVarSource``) or as a new volume (``ConfigDataVolumeSource``).
+5. API server is watched for modifications to the object.
+6. After a modification, actions 3, 4, and 5 are repeated.
+
+**Note**: Environment variables specified in ``ConfigData`` will override the environment variables
+with the same name found in the system if consumed as environment variables.
+
+## Use Cases
+
+#### Consuming ConfigData as mounted volumes
+
+``PodConfigurationVol`` is intended to be used as a mounted volume containing two files with their respective contents.
```

Comment:
I think this sentence could read a little smoother.  This is really an example of using a `ConfigData` with the volume plugin.  How about `example-volume-config` for the name instead of `PodConfigurationVol`.